api/git/service.go

```diff
@@ -0,0 +1,155 @@
+package git
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+var (
+	ErrIncorrectRepositoryURL = errors.New("Git repository could not be found, please ensure that the URL is correct.")
+	ErrAuthenticationFailure  = errors.New("Authentication failed, please ensure that the git credentials are correct.")
+	ErrRefNotFound            = errors.New("The target ref is not found")
+
+	REPOSITORY_CACHE_SIZE = 4
+	REPOSITORY_CACHE_TTL  = 30 * time.Minute
```

Comment:
Had a second thought on this value, i think we may need to keep it shorter, like 5-10mins.
The main use for a cached item would be to support searching, while a person sets their Stack. That's shouldn't take longer than 10mins. On the other hand if we'd keep it relatively large, all new changes that have been pushed won't appear for a user until a cache item would expire.